pkg/providers/vsphere/upgrade/virtualmachine/vm_schema_upgrade.go

```diff
@@ -427,31 +433,186 @@ func reconcileDevices(
+// schema upgrade. It maps disk devices from the vSphere VM
+// configuration to the VirtualMachine spec volumes, updating the
+// unit number, controller type, and controller bus number for each
+// PVC volume based on the disk UUID and controller information.
+// Once VMSharedDisks is enabled, a new VM's PVC is never needed to be
+// backfilled because it is expected that the PVCs to already have all the
+// placement info. The backfilling is only needed for existing VMs.
+func reconcileVirtualDisks(
 	ctx context.Context,
-	_ *vmopv1.VirtualMachine,
-	_ *vimtypes.VirtualDisk) {
+	k8sClient ctrlclient.Client,
+	vm *vmopv1.VirtualMachine,
+	ctrlKeyToCtrlStatus map[int32]vmopv1.VirtualControllerStatus,
+	diskDevices []*vimtypes.VirtualDisk) {
 
 	logger := pkglog.FromContextOrDefault(ctx)
-	logger.V(4).Info("Reconciling schema upgrade for VM disk")
+	logger.V(4).Info("Reconciling schema upgrade for VM disks")
+
+	if vm.Status.BiosUUID == "" {
+		logger.V(4).Info(
+			fmt.Sprintf(skipDisksReconciliationMsg, "BiosUUID not available"))
+		return
+	}
+
+	diskUUIDToDiskInfo := buildDiskUUIDToDiskInfoMap(diskDevices)
+	if len(diskUUIDToDiskInfo) == 0 {
+		logger.V(4).Info(fmt.Sprintf(
+			skipDisksReconciliationMsg, "no virtual disks attached"))
+		return
+	}
+
+	pvcNameToDiskUUID := buildPVCNameToDiskUUIDMap(ctx, k8sClient, vm)
+	if len(pvcNameToDiskUUID) == 0 {
+		logger.V(4).Info(fmt.Sprintf(
+			skipDisksReconciliationMsg, "no node attachments found"))
+		return
+	}
+
+	for i := range vm.Spec.Volumes {
+		vol := &vm.Spec.Volumes[i]
+
+		if vol.PersistentVolumeClaim == nil ||
+			vol.PersistentVolumeClaim.UnmanagedVolumeClaim != nil {
+			continue
+		}
+
+		reconcilePVCVolumePlacement(
+			ctx,
+			vol,
+			pvcNameToDiskUUID,
+			diskUUIDToDiskInfo,
+			ctrlKeyToCtrlStatus)
+	}
+}
+
+// buildDiskUUIDToDiskInfoMap builds a map of disk UUID to disk
+// device info by extracting from vSphere VM hardware.
+func buildDiskUUIDToDiskInfoMap(
+	diskDevices []*vimtypes.VirtualDisk) map[string]pkgutil.VirtualDiskInfo {
+
+	diskUUIDToDiskInfo := make(map[string]pkgutil.VirtualDiskInfo)
+	for _, disk := range diskDevices {
+		vdi := pkgutil.GetVirtualDiskInfo(disk)
+		if vdi.UUID != "" {
+			diskUUIDToDiskInfo[vdi.UUID] = vdi
+		}
+	}
+	return diskUUIDToDiskInfo
+}
+
+// buildPVCNameToDiskUUIDMap builds a map of PVC claim name to
+// disk UUID by querying CnsNodeVmAttachment objects.
+func buildPVCNameToDiskUUIDMap(
+	ctx context.Context,
+	k8sClient ctrlclient.Client,
+	vm *vmopv1.VirtualMachine) map[string]string {
+
+	attachments, err := pkgutil.GetAttachmentsForVM(ctx, k8sClient, vm)
+	if err != nil {
+		pkglog.FromContextOrDefault(ctx).V(4).Info(err.Error())
+		return nil
+	}
+
+	pvcNameToDiskUUID := make(map[string]string)
+	for _, attm := range attachments {
+		diskUUID := attm.Status.AttachmentMetadata[cnsv1alpha1.AttributeFirstClassDiskUUID]
+		pvcName := attm.Spec.VolumeName
+		if attm.Status.Attached && diskUUID != "" && pvcName != "" {
```

Comment:
What happens if this is attached after schema upgrade?

Reply:
Since PVC backfilling is not needed for a new VM, this is mostly for a brownfield VM with legacy CnsNodeVmAttachment. 

If a PVC spec is not backfilled during the schema upgrade, the mutation webhook will assign it to the next available slot which then becomes the SOT.  

Since the PVC is already tracked by the legacy CnsNodeVmAttachment, CnsNodeVMBatchAttachment will not be updated to reflect the update. Our newly added hardware device condition will show an error to indicate the mismatch. 

I am picturing this is a special case under the "Support mutating PVC settings" which we have a JIRA to keep track. I will add this case to it as well.

Reply:
Actually, we could ignore the Attached status here since we are checking directly against the attached virtual devices. This helps in the event that maybe the CnsNodeVmAttachment has not been updated yet.